src/pcnt.rs

```diff
@@ -0,0 +1,578 @@
+        lctrl_mode: PcntControlMode) -> Result<(), EspError> {
+        unsafe {
+            esp!(pcnt_set_mode(self.unit, channel.into(), pos_mode.into(), neg_mode.into(), hctrl_mode.into(), lctrl_mode.into()))
+        }
+    }
+
+    /// Add ISR handler for specified unit.
+    /// 
+    /// This ISR handler will be called from an ISR. So there is a stack
+    /// size limit (configurable as \"ISR stack size\" in menuconfig). This
+    /// limit is smaller compared to a global PCNT interrupt handler due
+    /// to the additional level of indirection.
+    /// 
+    /// # Safety
+    ///
+    /// Care should be taken not to call STD, libc or FreeRTOS APIs (except for a few allowed ones)
+    /// in the callback passed to this function, as it is executed in an ISR context.
+    /// 
+    /// @param callback Interrupt handler function.
+    /// 
+    /// returns
+    /// - ()
+    /// - EspError
+    #[cfg(feature = "alloc")]
+    pub unsafe fn subscribe(&self, callback: impl FnMut(u32) + Send + 'static) -> Result<(), EspError> {
+        enable_isr_service()?;
+
+        //self.unsubscribe();
+        let callback: Box<dyn FnMut(u32) + 'static> = Box::new(callback);
+        ISR_HANDLERS[self.unit as usize] = Some(callback);
+        esp!(pcnt_isr_handler_add(
+            self.unit,
+            Some(Self::handle_isr),
+            self.unit as *mut core::ffi::c_void,
+        ))?;
+        Ok(())
+    }
+
+    /// Remove ISR handler for specified unit.
+    /// 
+    /// returns
+    /// - ()
+    /// - EspError
+    #[cfg(feature = "alloc")]
+    pub fn unsubscribe(&self) -> Result<(), EspError> {
+        unsafe {
+            esp!(pcnt_isr_handler_remove(self.unit))?;
+            ISR_HANDLERS[self.unit as usize] = None;
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "alloc")]
+    unsafe extern "C" fn handle_isr(data: *mut core::ffi::c_void) {
+        let unit = data as pcnt_unit_t;
+        if let Some(f) = &mut ISR_HANDLERS[unit as usize] {
+            let mut value = 0u32;
+            esp!(pcnt_get_event_status(
+                unit,
+                &mut value as *mut u32
+            )).expect("failed to fetch event status!");
+            f(value);
+        }
+    }
+}
+
+impl Drop for PcntDriver<'_> {
+    fn drop(&mut self) {
+        let _ = self.counter_pause();
+        let _ = self.intr_disable();
+        unsafe {ISR_HANDLERS[self.unit as usize] = None};
+    }
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static ISR_SERVICE_ENABLED: core::sync::atomic::AtomicBool =
+    core::sync::atomic::AtomicBool::new(false);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static PCNT_CS: crate::task::CriticalSection = crate::task::CriticalSection::new();
+
+#[cfg(feature = "alloc")]
+fn enable_isr_service() -> Result<(), EspError> {
+    use core::sync::atomic::Ordering;
+
+    if !ISR_SERVICE_ENABLED.load(Ordering::SeqCst) {
+        let _ = PCNT_CS.enter();
```

Comment:
Replace `_` with `_cs`, or else it is not protected, as `_` vars are dropped immediately.